packages/openfeature-js/src/lib/types.ts

```diff
@@ -138,25 +199,63 @@ export interface ProviderOptions<T = unknown> {
   contextTransformer?: ContextTransformer<T>;
 }
 
-export interface Client extends HasHooks, Features {
+export type ProviderEvaluation<T> = {
+  value: T;
+  variant?: string;
+  reason: Reason;
+  errorCode?: ErrorCode;
+};
+
+export type FlagEvaluationDetails<T extends FlagValue> = {
+  flagKey: string;
+  executedHooks: ExecutedHooks;
+} & ProviderEvaluation<T>;
+
+export type ExecutedHooks = {
+  [P in keyof Omit<Hook<FlagValue>, 'name'>]-?: string[];
+};
+
+export enum Reason {
+  DISABLED = 'DISABLED',
+  SPLIT = 'SPLIT',
+  TARGETING_MATCH = 'TARGETING_MATCH',
+  DEFAULT = 'DEFAULT',
+  UNKNOWN = 'UNKNOWN',
+  ERROR = 'ERROR',
+}
+
+export enum ErrorCode {
+  PROVIDER_NOT_READY = 'PROVIDER_NOT_READY',
+  FLAG_NOT_FOUND = 'FLAG_NOT_FOUND',
+  PARSE_ERROR = 'PARSE_ERROR',
+  TYPE_MISMATCH = 'TYPE_MISMATCH',
+  GENERAL = 'GENERAL',
+}
```

Comment:
Here's another point for discussion. Are there are other Reasons or ErrorCodes we've clearly missed?

Reply:
Are we expecting providers to be able to extend these? I think that would provide more flexibility as things change. 

There are a few missing from what we current have in LaunchDarkly but some of these may be provider such as `FALLTHROUGH` https://docs.launchdarkly.com/sdk/concepts/evaluation-reasons

For Error codes we outline ours at: https://docs.launchdarkly.com/sdk/concepts/evaluation-reasons#error-conditions

Reply:
I guess there's no reason not to allow these to be extended, though I think that having `ErrorCodes` and `Reasons` beyond those defined in the spec might impact portability somewhat. Maybe that's an area we don't want to be to rigid about though.

Reply:
I see what you're saying. I think we should decide whether we want OpenFeature to define any error semantics or reasons of it's own then, or just some basic structures which underlying providers can populate with their own error codes and evaluation reasons. We could just make both of these string values, and then the provider could put whatever string makes sense into the data structure which would identify the reason/error, while not losing any fidelity. cc @beeme1mr

Reply:
I guess it depends on whether or not we think this information would be used to make runtime decisions. If it will, having consistency between providers would be useful. However, if it's just for troubleshooting/OTel, I would say we could just allow the provider to decide. Our spec could contain recommended reasons but not provide a hard requirement.

After typing this out, I think I'm leaning more towards keeping it general in OpenFeature and letting the providers decide. What do you think @toddbaert  and @InTheCloudDan?

Reply:
@beeme1mr that's a good point on runtime decision making. That may be more useful in the case of `error` than `evaluation reason`. In `error` you might want to alert if `FLAG_NOT_FOUND`.

Reply:
I think we can fairly say that the requirement to make runtime decisions based on error codes is sensible (though still probably not a very common use case) and so it might make sense to define those well, and map them from those in the provider. That said, pretty much all the SDKs don't throw exceptions, so our ability to actually propagate errors originating from the SDKs is quite limited. It seems somewhat less likely (and potentially an anti-pattern?) to make runtime decisions based on flag evaluation reason - that almost sounds like a hack to implement pre-requisite rules provided by some vendors.

I think perhaps we should make the reason a string for maximum flexibility, and keep the well-defined errors for now. It may be that later, some SDKs will support a "throw errors" mode that will give Openfeature more error info (errors would still be caught by OpenFeature, but more accurate information about the error could be logged).

Reply:
@InTheCloudDan @beeme1mr I think this issue is bigger than this PR. If it's OK with you, I will add an in-line comment around the `Reason` enum, open an new issue expressly for making a decision on this subject, and resolve this discussion for now, since it's not really the main issue of this PR.

Reply:
Let's continue this conversation here: https://github.com/open-feature/spec/issues/32#issuecomment-1101612389